sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala

```diff
@@ -796,4 +973,228 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
+      assert(get(result, "a", 3).get == 3)
+      assert(get(result, "a", 4).isEmpty)
+
+      provider.close()
+    }
+  }
+
+  protected def testSnapshotPartitionId(): Unit = {
+    withTempDir { tempDir =>
+      val inputData = MemoryStream[Int]
+      val df = inputData.toDF().limit(10)
+
+      testStream(df)(
+        StartStream(checkpointLocation = tempDir.getAbsolutePath),
+        AddData(inputData, 1, 2, 3, 4),
+        CheckLastBatch(1, 2, 3, 4)
+      )
+
+      val stateDf = spark.read.format("statestore")
+        .option(StateSourceOptions.SNAPSHOT_START_BATCH_ID, 0)
+        .option(StateSourceOptions.SNAPSHOT_PARTITION_ID, 0)
+        .option(StateSourceOptions.BATCH_ID, 0)
+        .load(tempDir.getAbsolutePath)
+
+      // should result in only one partition && should not throw error in planning stage
+      assert(stateDf.rdd.getNumPartitions == 1)
+
+      // should throw error when partition id is out of range
+      val stateDfError = spark.read.format("statestore")
+        .option(StateSourceOptions.SNAPSHOT_START_BATCH_ID, 0)
+        .option(
+          StateSourceOptions.SNAPSHOT_PARTITION_ID, 1)
+        .option(StateSourceOptions.BATCH_ID, 0)
+        .load(tempDir.getAbsolutePath)
+
+      val exc = intercept[StateStoreSnapshotPartitionNotFound] {
+        stateDfError.show()
+      }
+      assert(exc.getErrorClass === "CANNOT_LOAD_STATE_STORE.SNAPSHOT_PARTITION_ID_NOT_FOUND")
+    }
+  }
+
+  private def testSnapshotStateDfAgainstStateDf(resourceDir: File): Unit = {
+    val stateSnapshotDf = spark.read.format("statestore")
+      .option("snapshotPartitionId", 0)
+      .option("snapshotStartBatchId", 1)
+      .load(resourceDir.getAbsolutePath)
+
+    val stateDf = spark.read.format("statestore")
+      .load(resourceDir.getAbsolutePath)
+      .filter(col("partition_id") === 0)
+
+    checkAnswer(stateSnapshotDf, stateDf)
+  }
+
+  protected def testSnapshotOnLimitState(providerName: String): Unit = {
+    /** The golden files are generated by:
+    withSQLConf({
+      SQLConf.STREAMING_MAINTENANCE_INTERVAL.key -> "100"
+    }) {
+      val inputData = MemoryStream[(Int, Long)]
+      val query = inputData.toDF().limit(10)
+      testStream(query)(
+        StartStream(checkpointLocation = <...>),
+        AddData(inputData, (1, 1L), (2, 2L), (3, 3L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (4, 4L), (5, 5L), (6, 6L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (7, 7L), (8, 8L), (9, 9L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (10, 10L), (11, 11L), (12, 12L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) }
+      )
+    }
+     */
+    val resourceUri = this.getClass.getResource(
+      s"/structured-streaming/checkpoint-version-4.0.0/$providerName/limit/"
```

Comment:
I thought we were going to run against 3.5.1 and then run the query once to generate the operator metadata. Did we decide against that ?

Reply:
Strictly saying, the test about checkpoint with no operator metadata to create operator metadata should have been done in state metadata testing. If we don't have one, we'd better to have one, but no need to couple with this PR.